python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py

```diff
@@ -336,6 +392,82 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
         lstq = np.sqrt(s_assoc * s_cls)
         return lstq, s_assoc
 
+    def get_pat(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Calculate Panoptic Tracking (PAT) metric. https://arxiv.org/pdf/2109.03805.pdf
+       :return: (PAT, mean_PQ, mean_TQ).
+            PAT: <float64, 1>, PAT score over all classes.
+            mean_PQ: <float64, 1>, mean PQ scores over all classes.
+            mean_TQ: <float64, 1>, mean TQ score over all classes.
+        """
+        # First calculate for all classes
+        sq_all = self.pan_iou.astype(np.double) / np.maximum(self.pan_tp.astype(np.double), self.eps)
+        rq_all = self.pan_tp.astype(np.double) / np.maximum(
+            self.pan_tp.astype(np.double) + 0.5 * self.pan_fp.astype(np.double) + 0.5 * self.pan_fn.astype(np.double),
+            self.eps)
+        pq_all = sq_all * rq_all
+
+        # Then do the REAL mean (no ignored classes)
+        pq = pq_all[self.include].mean()
+
+        accumulate_tq = 0.0
+        accumlate_norm = 0
+
+        for seq in self.sequences:
+            preds = self.instance_preds[seq]
+            for cl in self.include:
+                cls_gts = self.instance_gts[seq][cl]
+                for gt_id, pr_ids in cls_gts.items():
+                    unique_pr_id, counts_pr_id = np.unique(pr_ids, return_counts=True)
+
+                    track_length = len(pr_ids)
+                    # void/stuff have instance value 1 due to the +1 in ln205 as well as unmatched gt is denoted by 1 
+                    # Thus we remove 1 from the prediction id list  
+                    unique_pr_id, counts_pr_id = unique_pr_id[unique_pr_id != 1], counts_pr_id[unique_pr_id != 1] 
```

Comment:
tks for the inline comment, sigh, this looks like a hack on top of another hack..

Reply:
Yup